transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapPacketLogger.java

```diff
@@ -0,0 +1,217 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.channel.epoll;
+
+import java.net.InetAddress;
+import java.util.Formatter;
+import java.util.Locale;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.logging.LogLevel;
+import io.netty.handler.logging.LoggingHandler;
+
+/**
+ * A {@link ChannelLogger} that logs packets sent or received via a {@link TunTapChannel}.
+ */
+public class TunTapPacketLogger extends LoggingHandler {
+
+    private boolean logPacketData = true;
+
+    public TunTapPacketLogger() {
+        super(LogLevel.INFO);
+    }
+
+    public TunTapPacketLogger(LogLevel level) {
+        super(level);
+    }
+
+    public TunTapPacketLogger(String name) {
+        super(name);
+    }
+
+    public TunTapPacketLogger(String name, LogLevel level) {
+        super(name, level);
+    }
+
+    public boolean logPacketData() {
+        return logPacketData;
+    }
+
+    public void setLogPacketData(boolean logPacketData) {
+        this.logPacketData = logPacketData;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof TunTapPacket) {
+            logPacket((TunTapPacket) msg, ((TunTapChannel) ctx.channel()).isTapChannel(), "TUN/TAP PACKET RECEIVED");
+            ctx.fireChannelRead(msg);
+        } else {
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (msg instanceof TunTapPacket) {
+            logPacket((TunTapPacket) msg, ((TunTapChannel) ctx.channel()).isTapChannel(), "TUN/TAP PACKET SENT");
+            ctx.write(msg, promise);
+        } else {
+            super.write(ctx, msg, promise);
+        }
+    }
+
+    private void logPacket(TunTapPacket packet, boolean isEtherFrame, String eventName) {
+        if (logger.isEnabled(internalLevel)) {
+            logger.log(internalLevel, formatPacket(packet, isEtherFrame, eventName, logPacketData));
+        }
+    }
+
+    public static String formatPacket(TunTapPacket packet, boolean isEtherFrame, String title, boolean logPacketData) {
+        StringBuilder out = new StringBuilder();
+        Formatter formatter = new Formatter(out, Locale.US);
+        ByteBuf packetData = packet.packetData();
+        int protocol = packet.protocol();
+        int ipPacketOffset = 0;
+
+        formatter.format("%s (protocol %s, len %d):\n",
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) %n should be used in place of \n to produce the platform-specific line separator. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2275)